src/scanpy/metrics/_common.py

```diff
@@ -15,19 +15,24 @@
 
 
 @singledispatch
-def _resolve_vals(val: NDArray | sparse.spmatrix) -> NDArray | sparse.csr_matrix:
-    return np.asarray(val)
+def _resolve_vals(
+    val: NDArray | sparse.spmatrix | DaskArray,
+) -> NDArray | sparse.csr_matrix | DaskArray:
+    msg = f"Unsupported type {type(val)}"
+    raise TypeError(msg)
```

Comment:
Does this type hint make sense? Why not `Any`?

Reply:
Two points here:
- the `@singledispatch`-decorated function is the one actually seen by typing. So either we also add `@overload` or give it the catch-all typing for all variants. (See https://github.com/python/mypy/issues/11727)
- typing a function that *might* not return is by pretending the exception case doesn’t exist (i.e. every function also has type `(Any) -> NoReturn`). Only functions that indeed never return should have type `(Any) -> NoReturn` or so.

So together, *if* typing would work as expected together with `@singledispatch` you’d be right: this branch should have type `(Any) -> NoReturn`.

Reply:
Also I’d generally ban `Any` as it’s infectious. A parameter should either have a type or `object` in case we don’t know anything. (`object`s can still be made into other types in `if isinstance(val, SomeType)` blocks)

Reply:
Thanks so much for the explanation.  Super interesting, had never thought of it in terms of "all functions do that" before, and had forgotten about `singledispatch` typing weirdness.  Thanks!